File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] ans = new int[n-k+1];
        List<Integer> list = new ArrayList<>();
        int i=0,j=0;
        while(j < n){
            while(!list.isEmpty() && list.get(list.size() - 1) < nums[j]){ // removing from the right side
                list.remove(list.size() - 1);
            }
            list.add(nums[j]);
            if(j - i + 1 < k){
                j++;
            }
            else if(j - i + 1 == k) {
                int x = list.get(0);
                ans[i] = x;
                if(x == nums[i]){
                    list.remove(0);
                }
                i++;
                j++;
            }
        }
        return ans;
    }
}

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        deque<int> dq;
        int n = nums.size();
        for(int i=0;i<n;i++){
            // removing the indices which are out of the bound
            while(!dq.empty() && i - dq.front() >= k){
                dq.pop_front();
            }
            // removing the elements which are smaller than the current element
            // and also maintain a strictly decreasing order
            while(!dq.empty() && nums[dq.back()] <= nums[i]){
                dq.pop_back();
            }
            dq.push_back(i);
            if(i >= k-1){
                ans.push_back(nums[dq.front()]);
            }
        }
        return ans;
    }
};